File: mobile/lib/features/watchlist/data/portfolio_audit.dart
/// Modelos de `GET/POST /api/v1/watchlist/audit` — la Auditoría de Portafolio por IA.
///
/// Espejan `app/schemas/portfolio_audit.py`. Mismo principio de parseo que la Ficha de
/// Inteligencia Profunda: **todo enum del backend se parsea con fallback, nunca con `throw`**. Un
/// sector nuevo del lado del servidor tiene que degradar ese dato puntual, no tirar abajo la
/// auditoría entera.
///
/// La aclaración que atraviesa todo este archivo: los porcentajes son **equiponderados por
/// cantidad de activos**, no por dinero invertido — la watchlist no guarda cantidades ni precio de
/// compra. Por eso `weightingBasis` viaja en el contrato y la UI lo dice explícito: "40% en
/// Tecnología" acá significa 4 de cada 10 activos seguidos, no 40 centavos de cada peso.
library;

import 'package:flutter/foundation.dart';

import '../../../core/data/data_availability.dart';

export '../../../core/data/data_availability.dart'
    show DataAvailability, availabilityFromWire;

/// Sector en el vocabulario del producto. El backend normaliza el del proveedor
/// (`Technology` -> `TECNOLOGIA`) para que la UI pueda agrupar y colorear por un vocabulario
/// cerrado.
///
/// `sinClasificar` es el fallback y también un valor legítimo del backend: un símbolo cuyo sector
/// no se pudo resolver aparece así en vez de desaparecer de la torta.
enum PortfolioSector {
  tecnologia,
  salud,
  serviciosFinancieros,
  consumoDiscrecional,
  consumoBasico,
  industria,
  energia,
  materiales,
  serviciosPublicos,
  bienesRaices,
  comunicaciones,
  cripto,
  sinClasificar,
}

PortfolioSector sectorFromWire(String? value) => switch (value) {
      'TECNOLOGIA' => PortfolioSector.tecnologia,
      'SALUD' => PortfolioSector.salud,
      'SERVICIOS_FINANCIEROS' => PortfolioSector.serviciosFinancieros,
      'CONSUMO_DISCRECIONAL' => PortfolioSector.consumoDiscrecional,
      'CONSUMO_BASICO' => PortfolioSector.consumoBasico,
      'INDUSTRIA' => PortfolioSector.industria,
      'ENERGIA' => PortfolioSector.energia,
      'MATERIALES' => PortfolioSector.materiales,
      'SERVICIOS_PUBLICOS' => PortfolioSector.serviciosPublicos,
      'BIENES_RAICES' => PortfolioSector.bienesRaices,
      'COMUNICACIONES' => PortfolioSector.comunicaciones,
      'CRIPTO' => PortfolioSector.cripto,
      _ => PortfolioSector.sinClasificar,
    };

/// Nivel de concentración. El backend lo calcula con umbrales explícitos sobre el peso del sector
/// dominante y el índice de Herfindahl, tomando el peor de los dos — no se lo pide al modelo, así
/// que la misma cartera da siempre el mismo veredicto.
enum RiskLevel { baja, moderada, alta, critica }

/// `moderada` como fallback y no `baja`: ante un valor que este cliente no conoce, la afirmación
/// prudente no es "tu cartera está bien".
RiskLevel riskLevelFromWire(String? value) => switch (value) {
      'BAJA' => RiskLevel.baja,
      'MODERADA' => RiskLevel.moderada,
      'ALTA' => RiskLevel.alta,
      'CRITICA' => RiskLevel.critica,
      _ => RiskLevel.moderada,
    };

/// De dónde sale una advertencia de correlación. Es parte del contrato y no un detalle interno:
/// una correlación medida sobre precios reales y una inferida de que dos activos comparten sector
/// tienen fuerza probatoria muy distinta, y la UI las muestra con chips distintos.
enum CorrelationBasis { priceHistory, sector }

/// `sector` como fallback: es la afirmación más débil de las dos, así que un valor desconocido
/// degrada la evidencia en vez de presentarla como medición.
CorrelationBasis correlationBasisFromWire(String? value) => switch (value) {
      'PRICE_HISTORY' => CorrelationBasis.priceHistory,
      _ => CorrelationBasis.sector,
    };

@immutable
class SectorAllocation {
  const SectorAllocation({
    required this.sector,
    required this.label,
    required this.weightPct,
    required this.tickerCount,
    required this.tickers,
  });

  factory SectorAllocation.fromJson(Map<String, dynamic> json) =>
      SectorAllocation(
        sector: sectorFromWire(json['sector'] as String?),
        // La etiqueta legible la manda el backend: si agrega un sector, la app vieja igual muestra
        // su nombre en castellano en vez de un código en mayúsculas.
        label: json['label'] as String? ?? '—',
        weightPct: (json['weight_pct'] as num?)?.toDouble() ?? 0,
        tickerCount: (json['ticker_count'] as num?)?.toInt() ?? 0,
        tickers: [
          for (final ticker in json['tickers'] as List? ?? []) ticker as String,
        ],
      );

  final PortfolioSector sector;
  final String label;
  final double weightPct;
  final int tickerCount;

  /// Qué activos concretos forman este sector. Viajan además del porcentaje porque un "40% en
  /// Tecnología" sin decir cuáles son no es accionable.
  final List<String> tickers;
}

@immutable
class ConcentrationRisk {
  const ConcentrationRisk({
    required this.level,
    required this.headline,
    required this.topSector,
    required this.topSectorLabel,
    required this.topSectorWeightPct,
    required this.distinctSectors,
    required this.herfindahlIndex,
    required this.notes,
  });

  factory ConcentrationRisk.fromJson(Map<String, dynamic> json) =>
      ConcentrationRisk(
        level: riskLevelFromWire(json['level'] as String?),
        headline: json['headline'] as String? ?? '',
        topSector: sectorFromWire(json['top_sector'] as String?),
        topSectorLabel: json['top_sector_label'] as String? ?? '—',
        topSectorWeightPct:
            (json['top_sector_weight_pct'] as num?)?.toDouble() ?? 0,
        distinctSectors: (json['distinct_sectors'] as num?)?.toInt() ?? 0,
        herfindahlIndex: (json['herfindahl_index'] as num?)?.toDouble() ?? 0,
        notes: [
          for (final note in json['notes'] as List? ?? []) note as String,
        ],
      );

  final RiskLevel level;

  /// Frase ya armada por el backend ("70% concentrado en Tecnología — riesgo alto"). Se muestra tal
  /// cual en vez de componerla acá para que diga lo mismo en la app, en el push y en la narrativa
  /// de la IA.
  final String headline;
  final PortfolioSector topSector;
  final String topSectorLabel;
  final double topSectorWeightPct;
  final int distinctSectors;

  /// Índice de Herfindahl normalizado a 0-1. Es lo que hace comparables dos carteras: "35% en
  /// Tech" dice poco si el resto está repartido en 8 sectores o en 2.
  final double herfindahlIndex;
  final List<String> notes;
}

@immutable
class CorrelationWarning {
  const CorrelationWarning({
    required this.tickers,
    required this.basis,
    required this.coefficient,
    required this.observations,
    required this.message,
  });

  factory CorrelationWarning.fromJson(Map<String, dynamic> json) =>
      CorrelationWarning(
        tickers: [
          for (final ticker in json['tickers'] as List? ?? []) ticker as String,
        ],
        basis: correlationBasisFromWire(json['basis'] as String?),
        coefficient: (json['coefficient'] as num?)?.toDouble(),
        observations: (json['observations'] as num?)?.toInt(),
        message: json['message'] as String? ?? '',
      );

  final List<String> tickers;
  final CorrelationBasis basis;

  /// `null` cuando la base es `sector`: el backend no inventa un número para una advertencia que no
  /// midió, y la UI no debe rellenarlo.
  final double? coefficient;
  final int? observations;
  final String message;

  bool get isMeasured => basis == CorrelationBasis.priceHistory;
}

@immutable
class DiversificationSuggestion {
  const DiversificationSuggestion({
    required this.sector,
    required this.label,
    required this.rationale,
  });

  factory DiversificationSuggestion.fromJson(Map<String, dynamic> json) =>
      DiversificationSuggestion(
        sector: sectorFromWire(json['sector'] as String?),
        label: json['label'] as String? ?? '—',
        rationale: json['rationale'] as String? ?? '',
      );

  final PortfolioSector sector;
  final String label;
  final String rationale;
}

@immutable
class PortfolioAudit {
  const PortfolioAudit({
    required this.generatedAt,
    required this.positionCount,
    required this.weightingBasis,
    required this.availability,
    required this.sectorAllocation,
    required this.sectorDataAvailable,
    required this.riskConcentration,
    required this.correlationWarnings,
    required this.correlationMeasured,
    required this.diversificationSuggestions,
    required this.aiSummary,
    required this.aiSummaryAvailable,
    required this.degradationReason,
    required this.servedFromCache,
  });

  factory PortfolioAudit.fromJson(Map<String, dynamic> json) => PortfolioAudit(
        generatedAt: DateTime.parse(json['generated_at'] as String),
        positionCount: (json['position_count'] as num?)?.toInt() ?? 0,
        weightingBasis:
            json['weighting_basis'] as String? ?? 'EQUAL_WEIGHT_BY_COUNT',
        availability: availabilityFromWire(json['availability'] as String?),
        sectorAllocation: [
          for (final entry in json['sector_allocation'] as List? ?? [])
            SectorAllocation.fromJson(entry as Map<String, dynamic>),
        ],
        sectorDataAvailable: json['sector_data_available'] as bool? ?? false,
        riskConcentration: json['risk_concentration'] == null
            ? null
            : ConcentrationRisk.fromJson(
                json['risk_concentration'] as Map<String, dynamic>,
              ),
        correlationWarnings: [
          for (final entry in json['correlation_warnings'] as List? ?? [])
            CorrelationWarning.fromJson(entry as Map<String, dynamic>),
        ],
        correlationMeasured: json['correlation_measured'] as bool? ?? false,
        diversificationSuggestions: [
          for (final entry
              in json['diversification_suggestions'] as List? ?? [])
            DiversificationSuggestion.fromJson(entry as Map<String, dynamic>),
        ],
        aiSummary: json['ai_summary'] as String?,
        aiSummaryAvailable: json['ai_summary_available'] as bool? ?? false,
        degradationReason: json['degradation_reason'] as String?,
        servedFromCache: json['served_from_cache'] as bool? ?? false,
      );

  final DateTime generatedAt;
  final int positionCount;

  /// Hoy siempre `EQUAL_WEIGHT_BY_COUNT`. Se guarda como string y no como enum porque el cliente no
  /// necesita ramificar sobre él —solo mostrar la aclaración correcta— y un enum obligaría a
  /// releasear la app el día que el backend agregue `MARKET_VALUE`.
  final String weightingBasis;
  final DataAvailability availability;
  final List<SectorAllocation> sectorAllocation;
  final bool sectorDataAvailable;
  final ConcentrationRisk? riskConcentration;
  final List<CorrelationWarning> correlationWarnings;

  /// `true` solo si al menos un PAR se pudo medir contra precios reales. Con `false`, la ausencia
  /// de advertencias no significa "los medimos y no correlacionan" sino "no los pudimos medir" — y
  /// la UI tiene que decir esa diferencia.
  final bool correlationMeasured;
  final List<DiversificationSuggestion> diversificationSuggestions;
  final String? aiSummary;
  final bool aiSummaryAvailable;
  final String? degradationReason;
  final bool servedFromCache;

  /// Sin activos no hay nada que auditar. Se distingue del resto de los estados degradados porque
  /// tiene una acción clara asociada ("agregá activos"), no un motivo técnico.
  bool get isEmpty => positionCount == 0;

  /// `true` cuando los pesos son cantidad de activos y no dinero invertido. Es lo que gobierna el
  /// texto de la aclaración; el día que el backend soporte posiciones reales, este getter devuelve
  /// `false` y la UI deja de aclararlo sin más cambios.
  bool get isEqualWeighted => weightingBasis == 'EQUAL_WEIGHT_BY_COUNT';
}

File: app/schemas/portfolio_audit.py
"""Schemas de `GET/POST /api/v1/watchlist/audit` — la Auditoría de Portafolio por IA.

Cuatro bloques que se calculan en código de forma determinística (distribución por sector,
concentración de riesgo, correlaciones y sugerencias de diversificación) más una narrativa que
redacta Gemini sobre esos cuatro. El reparto es deliberado: los números y los veredictos son
auditables y reproducibles, y el modelo solo los cuenta en prosa. Pedirle al LLM que calcule la
concentración abriría la puerta a que dos corridas con la misma cartera den veredictos distintos.

**Advertencia central del contrato**: la Watchlist no guarda cantidades ni precio de compra —
es una lista de seguimiento, no un portafolio con posiciones. Entonces la distribución es
EQUIPONDERADA POR CANTIDAD DE ACTIVOS, y eso viaja explícito en `weighting_basis`. Decir "70% en
Tecnología" sin aclararlo daría a entender que son 70 centavos de cada peso invertido, que es una
afirmación que este sistema no tiene datos para hacer.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field

# Se reutiliza el enum de la Ficha de Inteligencia Profunda en vez de declarar uno idéntico: es el
# mismo concepto (un bloque llegó completo, parcial o no llegó) y tener dos enums con los mismos
# tres valores obligaría al cliente a mapear entre ellos sin ninguna ganancia.
from app.schemas.intelligence import DataAvailability


class PortfolioSector(str, Enum):
    """Sector en el vocabulario del producto, normalizado desde el del proveedor de fundamentales
    (`Technology` -> `TECNOLOGIA`, ver `_SECTOR_TRANSLATIONS` en el servicio).

    Vocabulario cerrado y no el string crudo de FMP: la UI colorea y agrupa por sector, y un valor
    libre del proveedor (o su renombre) rompería el agrupado en silencio.

    `CRIPTO` no existe en la taxonomía GICS que usan los proveedores de acciones — se asigna en
    código por `asset_type`, sin preguntarle a nadie: una cripto no tiene sector empresario, y
    tratarla como `SIN_CLASIFICAR` esconde que es una clase de activo distinta, que es justo el
    dato que más importa en una auditoría de concentración.
    """

    TECNOLOGIA = "TECNOLOGIA"
    SALUD = "SALUD"
    SERVICIOS_FINANCIEROS = "SERVICIOS_FINANCIEROS"
    CONSUMO_DISCRECIONAL = "CONSUMO_DISCRECIONAL"
    CONSUMO_BASICO = "CONSUMO_BASICO"
    INDUSTRIA = "INDUSTRIA"
    ENERGIA = "ENERGIA"
    MATERIALES = "MATERIALES"
    SERVICIOS_PUBLICOS = "SERVICIOS_PUBLICOS"
    BIENES_RAICES = "BIENES_RAICES"
    COMUNICACIONES = "COMUNICACIONES"
    CRIPTO = "CRIPTO"
    SIN_CLASIFICAR = "SIN_CLASIFICAR"


# Etiqueta legible de cada sector. Viaja en la respuesta (`SectorAllocation.label`) para que el
# cliente no tenga que mantener su propia tabla de traducciones del enum — si se agrega un sector,
# la app vieja igual muestra su nombre en castellano en vez de un código en mayúsculas.
_SECTOR_LABELS: dict[PortfolioSector, str] = {
    PortfolioSector.TECNOLOGIA: "Tecnología",
    PortfolioSector.SALUD: "Salud",
    PortfolioSector.SERVICIOS_FINANCIEROS: "Servicios financieros",
    PortfolioSector.CONSUMO_DISCRECIONAL: "Consumo discrecional",
    PortfolioSector.CONSUMO_BASICO: "Consumo básico",
    PortfolioSector.INDUSTRIA: "Industria",
    PortfolioSector.ENERGIA: "Energía",
    PortfolioSector.MATERIALES: "Materiales",
    PortfolioSector.SERVICIOS_PUBLICOS: "Servicios públicos",
    PortfolioSector.BIENES_RAICES: "Bienes raíces",
    PortfolioSector.COMUNICACIONES: "Comunicaciones",
    PortfolioSector.CRIPTO: "Cripto",
    PortfolioSector.SIN_CLASIFICAR: "Sin clasificar",
}


def sector_label(sector: PortfolioSector) -> str:
    """Etiqueta legible de un sector. Nunca falla: un sector nuevo sin traducción cae en su propio
    código en vez de tirar un `KeyError` en el medio de una auditoría.
    """

    return _SECTOR_LABELS.get(sector, sector.value)


class WeightingBasis(str, Enum):
    """Sobre qué se calcularon los porcentajes.

    La Auditoría usa `EQUAL_WEIGHT_BY_COUNT` porque la watchlist no guarda cantidades: reparte por
    cantidad de activos. El Constructor de Portafolios sí conoce el capital de cada posición y usa
    `MARKET_VALUE`.

    Que el campo exista en las dos respuestas es lo que evita el error de leer un 40% contra el otro:
    "40% del capital" y "40% de los símbolos que sigo" son afirmaciones distintas sobre la misma
    cartera, y sin este campo un cliente tendría que deducir cuál está mirando.
    """

    EQUAL_WEIGHT_BY_COUNT = "EQUAL_WEIGHT_BY_COUNT"
    MARKET_VALUE = "MARKET_VALUE"


class RiskLevel(str, Enum):
    BAJA = "BAJA"
    MODERADA = "MODERADA"
    ALTA = "ALTA"
    CRITICA = "CRITICA"


class SectorAllocation(BaseModel):
    """Peso de un sector en la cartera, con los tickers que lo componen.

    Los tickers viajan además del porcentaje porque un "40% en Tecnología" sin decir cuáles son no
    es accionable: el usuario necesita saber qué activos concretos forman esa concentración.
    """

    model_config = ConfigDict(strict=True, extra="forbid", frozen=True)

    sector: PortfolioSector
    label: str
    weight_pct: float = Field(ge=0, le=100)
    ticker_count: int = Field(ge=0)
    tickers: list[str] = Field(default_factory=list)


class ConcentrationRisk(BaseModel):
    """Veredicto de concentración: nivel, titular legible y las señales que lo sostienen.

    `headline` viene armado del backend (ej. "70% concentrado en Tecnología — riesgo alto") en vez
    de dejar que el cliente lo componga: es la frase que el usuario va a leer primero, y tenerla
    acá la mantiene consistente entre la app, el push y la narrativa de la IA.
    """

    model_config = ConfigDict(strict=True, extra="forbid", frozen=True)

    level: RiskLevel
    headline: str
    top_sector: PortfolioSector
    top_sector_label: str
    top_sector_weight_pct: float = Field(ge=0, le=100)
    distinct_sectors: int = Field(ge=0)
    # Índice de Herfindahl-Hirschman sobre los pesos por sector, normalizado a 0-1. 1 es todo en un
    # solo sector; 1/N es el reparto perfecto entre N sectores. Se expone además del nivel porque
    # es la medida que hace comparables dos carteras: "35% en Tech" dice poco si el resto está
    # repartido en 8 sectores o en 2.
    herfindahl_index: float = Field(ge=0, le=1)
    notes: list[str] = Field(default_factory=list)


class CorrelationBasis(str, Enum):
    """De dónde sale la advertencia de correlación. Es parte del contrato y no un detalle interno:
    una correlación medida sobre precios reales y una inferida de que dos activos comparten sector
    tienen fuerza probatoria muy distinta, y el usuario merece saber cuál está leyendo.
    """

    PRICE_HISTORY = "PRICE_HISTORY"
    SECTOR = "SECTOR"


class CorrelationWarning(BaseModel):
    model_config = ConfigDict(strict=True, extra="forbid", frozen=True)

    tickers: list[str]
    basis: CorrelationBasis
    # `None` cuando la base es SECTOR: no se inventa un número para una advertencia que no se midió.
    coefficient: float | None = Field(default=None, ge=-1, le=1)
    observations: int | None = Field(default=None, ge=0)
    message: str


class DiversificationSuggestion(BaseModel):
    """Un sector ausente o subrepresentado que ayudaría a balancear la cartera.

    Se sugieren SECTORES y nunca tickers concretos, aunque el pedido admitía las dos cosas: "comprá
    X" es una recomendación de inversión personalizada, y este producto declara explícitamente que
    no las da (ver el disclaimer de la Ficha). Un sector es una observación sobre la forma de la
    cartera; un símbolo sería un consejo sobre qué hacer con el dinero.
    """

    model_config = ConfigDict(strict=True, extra="forbid", frozen=True)

    sector: PortfolioSector
    label: str
    rationale: str


class PortfolioAudit(BaseModel):
    """Respuesta de `GET/POST /api/v1/watchlist/audit`.

    Siempre 200 con estructura válida (salvo 401): una watchlist vacía, un catálogo sin sectores o
    un entorno sin credenciales de IA devuelven la auditoría con los bloques que se pudieron armar
    y `availability` + `degradation_reason` explicando el resto, en vez de un 4xx/5xx que el cliente
    tendría que traducir a una pantalla vacía.
    """

    model_config = ConfigDict(strict=True, extra="forbid", frozen=True)

    generated_at: datetime
    position_count: int = Field(ge=0)
    weighting_basis: WeightingBasis = WeightingBasis.EQUAL_WEIGHT_BY_COUNT

    availability: DataAvailability

    sector_allocation: list[SectorAllocation] = Field(default_factory=list)
    # `False` cuando ningún símbolo pudo clasificarse (sin FMP y sin sectores en el catálogo). La
    # distribución igual viaja, con todo en `SIN_CLASIFICAR`: es más honesto mostrar "no sabemos el
    # sector de estos 6 activos" que una torta vacía.
    sector_data_available: bool = True

    risk_concentration: ConcentrationRisk | None = None

    correlation_warnings: list[CorrelationWarning] = Field(default_factory=list)
    # `True` si al menos un PAR se pudo medir contra precios reales (no basta con haber bajado
    # histórico: hacen falta dos series con suficientes días en común). Con `False`, las
    # advertencias que haya son heurísticas por sector — y, más importante, la AUSENCIA de
    # advertencias no significa "los medimos y no correlacionan" sino "no los pudimos medir".
    correlation_measured: bool = False

    diversification_suggestions: list[DiversificationSuggestion] = Field(
        default_factory=list
    )

    ai_summary: str | None = None
    ai_summary_available: bool = False

    degradation_reason: str | None = None
    served_from_cache: bool = False
